WordPress/wordpress-playground
# [Blueprints] Login step – handle passwordless autologin via a PHP mu-plugin

Fixes #1706

## Motivation for the change, related issues

Handles auto login, e.g. in the `login` Blueprint step, using a platform-provided mu-plugin. This enables using the `login` step across all Playground runtimes without any of the issues with the previous implementation.

Before this PR, the `login` worked by making HTTP requests to the `/wp-login.php` endpoint. That approach had significant downsides:

* It only worked in web browsers
* It didn't support custom login mechanisms
* It required storing plaintext passwords in the Blueprint files
* It broke after a content import if the admin password changed
* It broke if WordPress displayed the [Site Admin Email Verification Screen](https://make.wordpress.org/core/2019/10/17/wordpress-5-3-admin-email-verification-screen/).

## Implementation details

To fix it and make the login step more robust login now uses a login PHP file that logs in the user with their credentials. 
The new script only requires a username, the password is now deprecated.

There are two ways to trigger the auto-login:

### The `PLAYGROUND_AUTO_LOGIN_AS_USER` constant

Used by the `login` Blueprint step does.

When the `PLAYGROUND_AUTO_LOGIN_AS_USER` constant is defined, this mu-plugin
will automatically log the user in on their first visit. The username is
the value of the constant.

On subsequent visits, the `playground_auto_login_happened` cookie will be
detected and the user will not be logged in. This means the "logout" feature
will work as expected.

### The `playground_force_auto_login_as_user` GET parameter

Used by the "login" button in various Playground runtimes.

When the `playground_force_auto_login_as_user` GET parameter is present,
this mu-plugin will automatically log in any logged out visitor. This will
happen every time they visit, not just on their first visit.
 
## Testing Instructions (or ideally a Blueprint)

- [Open this blueprint and confirm that login works](http://127.0.0.1:5400/website-server/?import-site=http%3A%2F%2F127.0.0.1%3A5400%2Fwebsite-server%2Ftest-fixtures%2Fsite-with-unpatched-gutenberg.zip&login=yes&url=%2Fwp-admin%2Fpost-new.php)
- Open Playground, log out, then navigate to `/?playground_force_auto_login_as_user=admin&` and confirm you were logged in again.

cc @brandonpayton – the `/?playground_force_auto_login_as_user=admin` triggers a redirect to add the trailing slash, as in `/?playground_force_auto_login_as_user=admin/`. Let's track that separately.
